batcher/batcher.mligo

```diff
@@ -149,40 +131,42 @@ let order_to_external (order: order) : external_order =
    Updates the current_batch if the time is valid but the new batch was not initialized. *)
-let deposit (external_order: external_order) (storage : storage) : result =
-  let current_batch = Batch.get_current_batch storage.batch_set in
-  let (last_order_number, batch_number) = match current_batch with
-                                             | None -> (0n, storage.batch_set.last_batch_number + 1n)
-                                             | Some cb -> (cb.last_order_number, storage.batch_set.current_batch_number) in
-  let order : order = external_to_order external_order last_order_number batch_number storage.valid_swaps in
-  let ticked_storage = tick_current_batches storage in
-  let current_time = Tezos.get_now () in
-  let updated_batch_set =
-    if Batch.should_open_new ticked_storage.batch_set current_time then
-      Batch.start_period order ticked_storage.batch_set current_time
-    else
-      try_to_append_order order ticked_storage.batch_set
-  in
-  let updated_user_orderbook = Userorderbook.push_open_order order.trader order storage.user_orderbook in
-  let updated_storage = { ticked_storage with batch_set = updated_batch_set; user_orderbook = updated_user_orderbook } in
-  let (tokens_transfer_op, storage_after_treasury_update) = Treasury.deposit order.trader order.swap.from updated_storage in
-  ([ tokens_transfer_op ], storage_after_treasury_update)
-
-let redeem (storage : storage) : result =
+let deposit (external_order: external_order) (old_storage : storage) : result =
+  let pair = Types.Utils.pair_of_external_swap external_order in
+  let ticked_storage = tick_current_batches pair old_storage in
+  let (current_batch_op, current_batch_set) = Batch.get_current_batch pair ticked_storage.batch_set in
+  match current_batch_op with
+  | None -> failwith Errors.no_open_batch
+  | Some current_batch-> let current_batch_number = current_batch.batch_number in
+                         let next_order_number = ticked_storage.last_order_number + 1n in
+                         let order : order = external_to_order external_order next_order_number current_batch_number ticked_storage.valid_swaps in
+                         let new_orderbook = Big_map.add next_order_number order ticked_storage.orderbook in
+                         let updated_volumes = Batch.update_volumes order current_batch in
+                         let updated_batches = Big_map.update current_batch_number (Some updated_volumes) current_batch_set.batches in
+                         let updated_batch_set = { current_batch_set with batches = updated_batches } in
+                         let updated_storage = { ticked_storage with batch_set = updated_batch_set; orderbook = new_orderbook; last_order_number = next_order_number; } in
+                         let tokens_transfer_op = Treasury.deposit order.trader order.swap.from in
+                         ([ tokens_transfer_op ], updated_storage)
+
+let redeem
+ (order_numbers: nat list)
```

Comment:
Currently, I don't understand the reasons why we need to redeem tokens in the corresponding `order_numbers` instead of redeeming all tokens.

Reply:
All orders are now in a big_map which I can't iterate over therefore I don't know which ones belong to a given user easily.  We could keep an user keyed orderbook as well but then we are back to having some form of list to hold the orders or order numbers.  

After speaking with @tomjack I am trying to implement a different type of storage to solve this so we won't need to reference the orders individually.

Reply:
LGTM.